### PR TITLE
Add the first composite type: Sum

### PR DESCRIPTION
**Pull request by rmjarvis**
This is finally the first composite type that I've been working up to.  Sum is probably a little simpler than Convolve, so I wanted to start with that.  But the other two I have in mind are Convolve and Transform, similar to the corresponding GalSIm object types. 

A Sum PSF is a model made from adding together multiple other PSF types.  Each component can have its own model and interpolation.  The parameters are stored in a list, so each component has its own numpy.array of parameters in the star.fit objects, each of which can be whatever length is required.

The initialization defaults to hsm for the first component and zero for the others, but the user can override this as desired.  I suspect for most use cases, the first component will describe most of the PSF patter, and the later components will fill in the bits that weren't well modeled earlier (e.g. the tree-ring residuals we've seen in DES), so I think this will be a reasonable default initialization.  Certainly we don't want multiple components all using the hsm initialization, since the sum would then be N x the PSF, which wouldn't be a good initial guess for the components.  I suspect that getting good initial guesses for various components will be a big part of making composite PSFs work well, so there may be more work needed eventually on this front.

### CODE REVIEW

File: piff/sumpsf.py [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+                        new_image -= current_models[kk][i].image
+                modified_star = Star(star.data.withNew(image=new_image), star.fit)
+                modified_stars.append(modified_star)
+
+            # Fit this component
+            new_stars, nremoved1 = comp.single_iteration(modified_stars, logger, convert_func)
+
+            # The fit components in new_stars are what we want, but the data parts are
+            # corrupted by subtracting the other components, which we don't want.
+            # Just copy over the fits.
```
Comment (theoschutt) [inline piff/sumpsf.py:197]: Something about this wording was very confusing. Possibly-too-wordy suggestion:

> new_stars now has the updated fit components, but the data parts are corrupted by subtracting the other components during fitting. Update the stars by copying the undisturbed data from the original stars and the updated fits from new_stars.

File: tests/test_sumpsf.py [hunk 2]
```diff
@@ -0,0 +1,507 @@
...
+    assert type(psf.components[1].model) is piff.PixelGrid
+    assert type(psf.components[1].interp) is piff.Polynomial
+
+    for i, star in enumerate(psf.stars):
+        print('star',i)
+        target = targets[i]
+        test_star = psf.drawStar(star)
+        # Note: There is no useful test to make of the component parameters.
+        #print('comp0: ',test_star.fit.params[0])
+        #print('comp1: ',test_star.fit.params[1])
```
Comment (theoschutt) [inline tests/test_sumpsf.py:314]: What do you mean here?

File: tests/test_sumpsf.py [hunk 3]
```diff
@@ -0,0 +1,507 @@
...
+
+        # Draw should produce something almost identical (modulo bounds).
+        test_image = psf.draw(x=test_star['x'], y=test_star['y'],
+                              stamp_size=config['input']['stamp_size'],
+                              flux=test_star.fit.flux, offset=test_star.fit.center/image.scale)
+        b = test_star.image.bounds & test_image.bounds
+        print('image max diff = ',np.max(np.abs(test_image[b].array - test_star.image[b].array)))
+        np.testing.assert_allclose(test_image[b].array, test_star.image[b].array, atol=1.e-8)
+
+    # Repeat with both components being PixelGrid mdoels.
```
Comment (theoschutt) [inline tests/test_sumpsf.py:274]: typo: mdoels -> models

**Review by theoschutt — APPROVED** [review]
Some in-line clarification questions, but otherwise looks good to me.

File: piff/sumpsf.py [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+
+        # Now initialize all the components
+        for k, comp in enumerate(self.components):
+            stars, nremoved1 = comp.initialize_params(stars, logger, default_init=default_init)
+            nremoved += nremoved1
+            # After the first one, set default_init to 'zero'
+            default_init='zero'
+
+        # If any components are degenerate, mark the sum as degenerate.
+        self.degenerate_points = any([comp.degenerate_points for comp in self.components])
```
[the remark this answers is not in the document]
Comment (rmjarvis) [reply piff/sumpsf.py:167]: That's correct.  The composition is of full PSF objects, so they don't interfere with each other.

The only thing this assignment does is to not do outlier rejection on the first iteration if any of the components use basis interpolation.  I'm not sure how important that is, but it seemed sensible.

File: piff/sumpsf.py [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+        # Fit each component in order
+        for k, comp in enumerate(self.components):
+            logger.info("Starting work on component %d (%s)", k, comp._type_name)
+            # Update stars to subtract current fit from other components.
+            modified_stars = []
+            for i, star in enumerate(stars):
+                new_image = star.image.copy()
+                for kk in range(len(self.components)):
+                    if kk != k:
+                        new_image -= current_models[kk][i].image
```
[the remark this answers is not in the document]
Comment (rmjarvis) [reply piff/sumpsf.py:188]: That's right.  We always subtract off the current best model of all other components, and fit the component being worked on to the residual.
I think this should be pretty stable when applying some tweak to a primary model that is pretty good at describing most of the pattern, but has a residual with a very different interpolation pattern.  E.g. maybe we add on a PixelGrid with much larger extent, but no interpolation, so it just captures the average large-scale residuals around a standard basis-interpolated model over a smaller footprint.  None of my unit tests tried this out, but that's what I mostly have in mind for how the SumPSF would get used in practice.

File: tests/test_sumpsf.py (hunk 2 above)
Comment (rmjarvis) [reply tests/test_sumpsf.py:314]: Usually I had been testing the fitted parameters to some known true values.  But for this test, both models are PixelGrid, so the parameters aren't anything I know a priori.  The only thing to test is to use the models to draw images of the total PSF model and see if it matches the data acceptably.

**Comment by rmjarvis** [comment]
@jmeyers314 @aaronroodman 
Would either of you like to take a look at this before merging?  I'll be doing Convolution next if you're rather wait for that as a more realistic composite model.

**Review by jmeyers314 — APPROVED** [review]
Hey Mike.  Code all looks good.  I'm wondering if there's room for another practical test though - something like Kolmogorov-to-capture-bulk + PixelGrid-to-mop-up-details, maybe with very different spatial scales?

**Comment by rmjarvis** [comment]
I'm game.  Thoughts on what to use for the truth?  Maybe a GalSim atmospheric screen and then use GP for the interpolation of Kolmogorov for the primary component.  Would that have a usefully interesting residual to mop up?  Or would I need to include some optical component to give it something that wouldn't be well described by Kolmogorov?

**Comment by jmeyers314** [comment]
I'm not sure.  If you keep the simulated exposure time short enough I'm sure an atmsopheric screen true psf would have interesting residuals to mop up.

You might be able to get away with something simpler though, e.g., truth = Moffat with quickly spatially varying beta maybe?
Definitely bonus points though if you get this to work with a GP.

**Comment by rmjarvis** [comment]
Hi Josh, I finally got back to this after 8 months of letting your very good suggestion sit unaddressed.

I added a test using Atmospheric + Optical screens for the PSF.  The model is a Kolmogorov PSF with GP interpolation + a PixelGrid with just a Mean interpolation.

If I don't include the PixelGrid part, here are the images of the residuals (made by the new unit test when running from main):
test_atmpsf_stars2
You can see that there is a roughly consistent pattern to the residuals.  Mostly because the optical screen aberrations were constant over the CCD, so they produces a pretty coherent pattern.

When I include the PixelGrid as a second component in the Sum, here are the residuals:
test_atmpsf_stars

This seems pretty good to me and shows a more realistic use case for how the Sum PSF might be used in practice.

**Pull request merged**